channel settings: Add a "Reset to default notifications" icon in settings.

saubhagya-patel opened the pull request:
This PR adds a "Reset to default notifications" icon in the Notifications menu, to the right of the channel name, for which a user has selected custom notifications and a confirmation dialog to confirm this action.


<!-- Describe your pull request here.-->


Note from the [issue](https://github.com/zulip/zulip/issues/19860):
>  In the notifications menu. The button should be a trash can icon to the right of the stream name, with a tooltip: "Reset stream to default notifications". Clicking on it should remove the line for the line for the stream.

Fixes: #19860.
[CZO](https://chat.zulip.org) thread: [#design > reset channel notifications to default icon](https://chat.zulip.org/#narrow/channel/101-design/topic/reset.20channel.20notifications.20to.20default.20icon).

<br/>


<details>
<summary><code>git grep</code> of the introduced class <code>reset_stream_notifications</code> for the icon.
</summary>

**Before:**
```bash
$ git grep reset_stream_notifications

```

**After:**
```bash
$ git grep reset_stream_notifications
web/src/settings_notifications.ts:import render_confirm_reset_stream_notifications from "../templates/confirm_dialog/confirm_reset_stream_notifications.hbs";
web/src/settings_notifications.ts:export function do_reset_stream_notifications(elem: HTMLElement, sub: StreamSubscription): void {
web/src/settings_notifications.ts:function reset_stream_notifications(elem: HTMLElement): void {
web/src/settings_notifications.ts:    const html_body = render_confirm_reset_stream_notifications({sub});
web/src/settings_notifications.ts:            do_reset_stream_notifications(elem, sub);
web/src/settings_notifications.ts:        "#stream-specific-notify-table .reset_stream_notifications",
web/src/settings_notifications.ts:            reset_stream_notifications(this);
web/src/stream_edit.ts:            settings_notifications.do_reset_stream_notifications(this, sub);
web/src/tippyjs.ts:            "#stream-specific-notify-table .reset_stream_notifications",
web/styles/settings.css:    .reset_stream_notifications {
web/styles/settings.css:    .reset_stream_notifications:hover {
web/styles/settings.css:            .reset_stream_notifications {
web/templates/settings/stream_specific_notification_row.hbs:        <i class="zulip-icon icon-button-neutral zulip-icon-reset reset_stream_notifications" role="button" tabindex="0" data-tippy-content="{{t 'Reset to default notifications' }}" aria-label="{{t 'Reset to default notifications' }}"></i>
```

</details>

<br/>


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
  <summary>Added "Reset to default notifications" icon in the Notifications menu.</summary>
  
  - Light mode:
    
  ![image](https://github.com/user-attachments/assets/b2a7870d-212c-4ef4-9b37-3cf2cbfed567)
  ![image](https://github.com/user-attachments/assets/8e5be74b-8cd1-4f47-b192-80debc7ee703)

  - Dark mode
    
  ![image](https://github.com/user-attachments/assets/1957a7ec-da47-4862-be20-16b407152368)
  ![image](https://github.com/user-attachments/assets/72e6d107-1040-4c97-8d4d-657b13c1cc14)

</details>


<details>
<summary>Tooltip on the reset to default notifications icon.</summary>
    
  | Light mode | Dark mode|
  |--------------|-------------|
  | ![image](https://github.com/user-attachments/assets/622d667d-599c-4dfb-9bac-cc8a90fa09a4) | ![image](https://github.com/user-attachments/assets/d7589bd5-ef22-46f7-8b02-7cca732d4c36) |

</details>

<details>
<summary>Example behavior of the reset to default notifications icon.</summary>

  ![reset icon example behavior](https://github.com/user-attachments/assets/e0ec4891-b4f3-4ac0-b730-5d63a36db3b1)

</details>


---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
You can mark this PR as fixing #19860, as it fixes the only remaining part.

alya commented:
Cool! Let's make the trash can have the same color and hover behavior as the `?` link on Topic notifications below.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes are made:
- Following [this](https://github.com/zulip/zulip/pull/32923#issuecomment-2573964516) comment, I have updated the [PR description](https://github.com/zulip/zulip/pull/32923#issue-2771108664) and [commit message](https://github.com/zulip/zulip/pull/32923/commits/1d23c68c4ef9f4a1888c3d791c7c72a360b4037d) to mark this PR as fixing the issue [#19860](https://github.com/zulip/zulip/issues/19860).

- The trash can icon now has the same color and hover effects as the help link (`?`) icon as mentioned in [this](https://github.com/zulip/zulip/pull/32923#issuecomment-2573964516) comment.

- **Visual changes:**

  <details>
  <summary>Updated color of trash can icon.</summary>
  
  - Light Mode

    **Without unmute icon:**
    ![image](https://github.com/user-attachments/assets/f9a97df9-3fef-4ca6-a06c-7d771f1d7fd3)

    **With unmute icon:**
    ![image](https://github.com/user-attachments/assets/cf4baaf8-ef20-46bc-b26b-8cb0222ce931)

  - Dark Mode

    **Without unmute icon:**
    ![image](https://github.com/user-attachments/assets/8706a82d-851f-4585-88bc-d4c1c7881b71)

    **With unmute icon:**
    ![image](https://github.com/user-attachments/assets/10bf7e9d-d32b-43d4-9814-197a228f25bb)

  </details>
  
  <details>
  <summary>Updated hover behavior of trash can icon.</summary>
  
  I am unsure whether the hover behavior is demonstrated properly.
  
  **Light mode:**
  ![reset notif light mode hover](https://github.com/user-attachments/assets/c3aa02f9-36f2-4a48-9d38-0f94cbf580fb)

  **Dark mode:**
  ![reset notif dark mode hover](https://github.com/user-attachments/assets/d8f491d3-12e5-46dd-9235-dd414fa9278f)

  </details>
  
@alya, this PR is ready for review. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

alya commented:
@sahil839 Could you please take a look at this one when you have time? Not a 10.0 prirority.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes are made:
- The unmute stream icon (`unmute_stream`) now follows the same styles as the help link (`?`) icon. I have made this change in a separate commit following [this](https://github.com/zulip/zulip/pull/32923#issuecomment-2576163798) comment.

- **Visual changes:**

  <details>
  <summary>Updated unmute stream icon.</summary>

  **Light mode:**
  ![image](https://github.com/user-attachments/assets/b8786d5a-d82e-4c80-b7ca-fea298913ce9)

  **Dark mode:**
  ![image](https://github.com/user-attachments/assets/3e560f55-fa40-4eae-8d11-6d67e3f2d365)

  </details>
  
  <details>
  <summary>Hover behavior of the unmute stream icon.</summary>
  
  **Light mode:**
  ![unmute icon hover light](https://github.com/user-attachments/assets/6d327345-b3cf-475a-8c68-47104dd47481)

  **Dark mode:**
  ![unmute icon hover dark](https://github.com/user-attachments/assets/4db9c9ca-4749-4295-b901-7ed18766033b)

  </details>
 
Thank you.

saubhagya-patel commented on web/src/settings_notifications.ts:
```diff
@@ -171,6 +178,41 @@ function stream_notification_setting_changed(target: HTMLInputElement): void {
+    )) {
+        data.push({
+            stream_id: sub.stream_id,
+            property: settings_labels_schema.parse(per_stream_setting_name),
```
Sure, updated it in the following push.

saubhagya-patel commented:
Following the review received, I have updated the PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/32923#discussion_r2009941839) comment, `settings_labels_schema` being used to parse the notification settings of a stream is replaced to use `stream_specific_notification_settings_schema`.

- Visual changes: _None_

@sahil839, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

sahil839 commented on web/templates/confirm_dialog/confirm_reset_stream_notifications.hbs:
```diff
@@ -0,0 +1,6 @@
+<p>
+    {{#tr}}
+    Are you sure you want to reset notifications for <z-stream></z-stream> ?
+    {{#*inline "z-stream"}}{{> ../inline_decorated_stream_name stream=sub show_colored_icon=true}}{{/inline}}
```
I think we should not show colored icon here. We just show black icons in at least one more modal I guess.

sahil839 commented on web/src/settings_notifications.ts:
```diff
@@ -171,6 +173,43 @@ function stream_notification_setting_changed(target: HTMLInputElement): void {
+
+    stream_settings_api.bulk_set_stream_property(
+        data,
+        $(elem).closest(".subsection-parent").find(".alert-notification"),
```
Instead of showing saving indicator, maybe we can just keep the dialog open and show the loading indicator in the button itself. And can also show error, if any, in the modal itself.

saubhagya-patel replied:
@sahil839, should I wait for @alya's review or should I go ahead and implement this?

sahil839 reviewed (commented):
Looks good to me overall. Have a couple of question regarding the behavior, so would be good to get feedback from @alya on that before making any changes.

saubhagya-patel commented:
@alya, could you please provide some feedback on these comments:
- Regarding showing colored icon in the reset notification modal - [comment](https://github.com/zulip/zulip/pull/32923#discussion_r2011489704).
- Regarding showing the error in the modal if any: [comment](https://github.com/zulip/zulip/pull/32923#discussion_r2011492120).

Thank you.

alya commented:
I think at this point we should be using modern Zulip buttons both for the muted indicator and the reset button. Danger icon style would be appropriate for the reset.

alya commented:
Can you post in #design to see if folks think we need a confirmation dialog? Playing with the PR, I'm kind of feeling like it's an unnecessary extra step.

alya commented:
If we get rid of the modal, replacing the trash can with a loading spinner might work well.

saubhagya-patel commented:
`Alya Abbott` [said](https://github.com/zulip/zulip/pull/32923#issuecomment-2782183202):
> Can you post in #design to see if folks think we need a confirmation dialog? Playing with the PR, I'm kind of feeling like it's an unnecessary extra step.

Sure, started a thread here: [#design > reset channel notifications to default icon](https://chat.zulip.org/#narrow/channel/101-design/topic/reset.20channel.20notifications.20to.20default.20icon)

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/32923#issuecomment-2782179380) comment, the `unmute_stream` icon is updated to use the modern Zulip mute icon(`zulip-icon-mute-new`), the changes are added in a prep commit: [settings: Update unmute icon in Personal settings > Notifications.](https://github.com/zulip/zulip/pull/32923/commits/08351ff5503198c9525c83fff5e1324644753f1d).

- For the comments: [[1]](https://github.com/zulip/zulip/pull/32923#issuecomment-2782183202), [[2]](https://github.com/zulip/zulip/pull/32923#issuecomment-2782186620), a thread: [#design > reset channel notifications to default icon](https://chat.zulip.org/#narrow/channel/101-design/topic/reset.20channel.20notifications.20to.20default.20icon) was started. Based on this thread, the following updates are made:
  - No changes are made to the confirmation dialog, which appears when a user clicks on the reset to default notifications icon.
  - The reset icon is changed from an `x` to a reset icon: https://lucide.dev/icons/rotate-ccw.
  - The reset icon appears when a user hovers over the cell of a particular channel.

- Following [this](https://github.com/zulip/zulip/pull/32923#issuecomment-2782184800) comment, the tooltip content is updated from "Reset channel to default notifications" -> "Reset to default notifications".

- <details>
  <summary>Visual changes</summary>
  
  - Added icon in the Personal Notifications settings:
  
    - Light mode:
    
    ![Screenshot 2025-05-18 002945](https://github.com/user-attachments/assets/17862b75-8f93-4e1b-be62-24b7d1c7efec)
    ![Screenshot 2025-05-18 002959](https://github.com/user-attachments/assets/d151ff30-f829-49ee-a7b6-47774587f67d)

    - Dark mode
    
    ![Screenshot 2025-05-18 002729](https://github.com/user-attachments/assets/df5d165b-ac4d-4e8b-90a8-deb46889da49)
    ![Screenshot 2025-05-18 002745](https://github.com/user-attachments/assets/275204c3-f878-4191-b3df-02d2c9187604)


  
  - Tooltip on the reset notifications icon:
    
  | Light mode | Dark mode|
  |--------------|-------------|
  | ![Screenshot 2025-05-18 003159](https://github.com/user-attachments/assets/7c339080-1fb4-4278-b58f-15515733952d) | ![Screenshot 2025-05-18 003243](https://github.com/user-attachments/assets/231f6280-7434-44aa-a168-d55eaf33c20c) |
  
  - Example behavior of the reset to default notifications icon:
  ![reset icon example behavior](https://github.com/user-attachments/assets/99ff6e3f-91c7-4e94-bd5c-f70ca596ac87)
  
  </details>


@alya, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

saubhagya-patel commented:
Following [this](https://chat.zulip.org/#narrow/channel/101-design/topic/reset.20channel.20notifications.20to.20default.20icon/near/2176841) message in the thread [#design > reset channel notifications to default icon](https://chat.zulip.org/#narrow/channel/101-design/topic/reset.20channel.20notifications.20to.20default.20icon), I have updated this PR. The following changes are made:

- This reset icon is updated:
  
  | Before | After |
  |--------------|-------------|
  | ![image](https://github.com/user-attachments/assets/de75f0f9-a1b5-4ccf-8d92-e33487d701c4) <br/> https://lucide.dev/icons/rotate-ccw | ![image](https://github.com/user-attachments/assets/cf86246e-453d-4d3b-a9ed-c9b7ff447390) <br/> https://lucide.dev/icons/undo-2 |
  

- <details>
  <summary>Visual changes</summary>
  
  - Added icon in the Personal Notifications settings:
  
    - Light mode:
    
    ![image](https://github.com/user-attachments/assets/b2a7870d-212c-4ef4-9b37-3cf2cbfed567)
    ![image](https://github.com/user-attachments/assets/8e5be74b-8cd1-4f47-b192-80debc7ee703)

    - Dark mode
    
    ![image](https://github.com/user-attachments/assets/1957a7ec-da47-4862-be20-16b407152368)
    ![image](https://github.com/user-attachments/assets/72e6d107-1040-4c97-8d4d-657b13c1cc14)

  
  - Tooltip on the reset notifications icon:
    
  | Light mode | Dark mode|
  |--------------|-------------|
  | ![image](https://github.com/user-attachments/assets/622d667d-599c-4dfb-9bac-cc8a90fa09a4) | ![image](https://github.com/user-attachments/assets/d7589bd5-ef22-46f7-8b02-7cca732d4c36) |
  
  - Example behavior of the reset to default notifications icon:
  ![reset icon example behavior](https://github.com/user-attachments/assets/e0ec4891-b4f3-4ac0-b730-5d63a36db3b1)
  
  </details>

Thank you.

saubhagya-patel commented on web/templates/settings/stream_specific_notification_row.hbs:
```diff
@@ -6,8 +6,7 @@
         {{stream.stream_name}}
-        <i class="zulip-icon zulip-icon-mute unmute_stream" data-tippy-content="{{t 'Unmute' }}"
-          {{#unless muted}}style="display: none;"{{/unless}}></i>
+        <i class="zulip-icon icon-button-brand zulip-icon-mute-new unmute_stream" role="button" tabindex="0" data-tippy-content="{{t 'Unmute' }}" aria-label="{{t 'Unmute' }}" {{#unless muted}}style="display: none;"{{/unless}}></i>
```
Sure, started a thread here: [#frontend > rename &#96;zulip-icon-mute-new&#96; icon to &#96;zulip-icon-mute&#96;](https://chat.zulip.org/#narrow/channel/6-frontend/topic/rename.20.60zulip-icon-mute-new.60.20icon.20to.20.60zulip-icon-mute.60/with/2178058).

timabbott commented:
Is there anything else we need to do after merging that icon PR?

alya commented:
Oh, and:

4. Looks like the muted channel marker has disappeared?

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/32923#issuecomment-2950149341) comment, the following changes are made:
  - Adjusted delay for reset icon.
  - Removed extra space between channel name and `?` in the confirmation modal to reset channel notifications.
  - Updated spacing for the channel privacy icon type, so that it doesn't vary with font size.

- Following [this](https://github.com/zulip/zulip/pull/32923#issuecomment-2950389755) comment, for the disappeared channel marker, this was because of #34701, not sure why GitHub didn't flag for merge conflicts.

- Please note that I have also adjusted the tooltip delay for the unmute channel icon so that the tooltip delay remains uniform.

- <details>
  <summary> Visual changes </summary>
  
  - Removed extra space between channel name and `?` in the confirmation modal to reset channel notifications.
  
  | Before | After |
  |--------------|-------------|
  | ![image](https://github.com/user-attachments/assets/8b213f63-b5f0-45c6-be38-a0919ffd7bc5) | ![image](https://github.com/user-attachments/assets/070a3738-2574-45a7-bb8e-ba9803fed4b6) |

  - Updated spacing for the channel privacy icon type, so that it doesn't vary with font size.
    | Font size | Modal |
    |--------------|-------------|
    | 12px | ![image](https://github.com/user-attachments/assets/bf23bdc0-03de-48a5-98bc-5d928edc6781) |
    | 16px | ![image](https://github.com/user-attachments/assets/89dec1e8-dc5e-4554-9131-5a2b50b5e5b2) |
    | 20px | ![image](https://github.com/user-attachments/assets/76ee62ca-9efc-44bc-8af3-ef1b90c92d1f) |

</details>

@alya, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

alya commented:
The next step is for @sahil839 to re-review.

sahil839 commented:
The channel privacy icon alignment looks off, but I can see that in main as well, so not related to this PR.

Looks good other than that.

timabbott commented:
Merged, thanks @saubhagya-patel and @sahil839! @saubhagya-patel can you make sure we're tracking that misaligned privacy icon, perhaps via a new thread in #issues if it's not been reported yet elsewhere?

saubhagya-patel commented:
> @saubhagya-patel can you make sure we're tracking that misaligned privacy icon, perhaps via a new thread in #issues if it's not been reported yet elsewhere?

Sure, started a thread here: [#issues > misaligned channel privacy icon in modals](https://chat.zulip.org/#narrow/channel/9-issues/topic/misaligned.20channel.20privacy.20icon.20in.20modals/with/2192394).

sahil839 commented:
I meant the icons in stream rows of the notifications table.

saubhagya-patel commented:
> I meant the icons in stream rows of the notifications table.

Ohh, my bad, I've updated the thread here: [#issues > misaligned channel privacy icon in modal, notification table @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/misaligned.20channel.20privacy.20icon.20in.20modal.2C.20notification.20table/near/2192409).
